#### A. Segmenting Customers Based on Financial Backgrounds, Poverty Levels, Spending Habits, and Income Levels
- **Best Algorithm: Clustering**  
  This is an **unsupervised learning problem**, where the goal is to segment customers into different groups based on their financial attributes. Clustering algorithms like **K-Means** or **Hierarchical Clustering** are well-suited for grouping similar customers without any labeled data.

#### B. Automatically Recognizing Bus from Car Using a Camera
- **Best Algorithm: Classification**  
  This scenario involves distinguishing between two categories—bus and car—which makes it a **classification problem**. Algorithms like **Logistic Regression**, **Decision Trees**, or **Convolutional Neural Networks (CNNs)** can be used to classify the images captured by the camera.

#### C. Automatically Recognize Friends Without Labeling on Facebook
- **Best Algorithm: Unsupervised Feature Extraction**  
  Since the goal is to recognize friends without any labeled data, **unsupervised feature extraction** is the best choice. Techniques like **Principal Component Analysis (PCA)** or clustering-based methods can be used to identify unique features of individuals in images without labels.

#### D. Identifying Accomplices of a Thief Based on Call Records
- **Best Algorithm: Association Rule Mining**  
  This scenario requires discovering relationships between individuals based on call records, which is typical for **association rule mining**. Algorithms like the **Apriori** or **FP-Growth** are used to find frequent associations or relationships between entities, making it useful for identifying accomplices.

#### E. Predicting Future Prices of Stocks Based on Volume of Sales and Price Differences
- **Best Algorithm: Regression**  
  Since the task is to predict a continuous numerical value (future stock prices), **regression** is the best approach. Techniques like **Linear Regression**, **Random Forest Regression**, or more advanced methods like **LSTM (Long Short-Term Memory)** for time-series forecasting are suitable for this problem.


In [36]:
from sklearn.model_selection import train_test_split,cross_val_score
from sklearn.linear_model import LinearRegression, SGDRegressor
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.metrics import mean_squared_error, r2_score
import optuna
import numpy as np
import pandas as pd
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import Ridge
import numpy as np
import joblib

In [ ]:
data = pd.read_csv('data/boston.csv')

In [4]:
print(data.columns)

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PT', 'B', 'LSTAT', 'MV'],
      dtype='object')


In [5]:
print(data.head)

<bound method NDFrame.head of         CRIM    ZN  INDUS  CHAS    NOX     RM        AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.199997  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.900002  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.099998  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.799999  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.200001  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...        ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.099998  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.699997  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.000000  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.300003  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.800003  2.5050    1  273   

            PT           B  LSTAT

In [6]:
print(data.info)

<bound method DataFrame.info of         CRIM    ZN  INDUS  CHAS    NOX     RM        AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.199997  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.900002  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.099998  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.799999  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.200001  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...        ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.099998  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.699997  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.000000  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.300003  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.800003  2.5050    1  273   

            PT           B  LST

In [7]:
print(data.describe)

<bound method NDFrame.describe of         CRIM    ZN  INDUS  CHAS    NOX     RM        AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.199997  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.900002  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.099998  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.799999  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.200001  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...        ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.099998  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.699997  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.000000  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.300003  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.800003  2.5050    1  273   

            PT           B  L

In [8]:
data.isnull().sum()

CRIM     0
ZN       0
INDUS    0
CHAS     0
NOX      0
RM       0
AGE      0
DIS      0
RAD      0
TAX      0
PT       0
B        0
LSTAT    0
MV       0
dtype: int64

In [9]:
for col in data.columns:
    print(col,len(data[col].unique()))
    print()
    

CRIM 504

ZN 26

INDUS 76

CHAS 2

NOX 81

RM 446

AGE 356

DIS 412

RAD 9

TAX 66

PT 46

B 357

LSTAT 455

MV 229



Features: The 13 features are numerical and describe various characteristics of the Boston housing market and neighborhoods, such as:

    - CRIM: Per capita crime rate by town (504 unique)
    - ZN: Proportion of residential land zoned for lots over 25,000 sq. ft. (26 unique)
    - INDUS: Proportion of non-retail business acres per town (76 unique)
    - CHAS: Charles River dummy variable (1 if tract bounds river; 0 otherwise) (2 unique) (Only Discrete column)
    - NOX: Nitric oxide concentration (parts per 10 million) (81 unique)
    - RM: Average number of rooms per dwelling (446 unique)
    - AGE: Proportion of owner-occupied units built before 1940 (356 unique)
    - DIS: Weighted distances to five Boston employment centers (412 unique)
    - RAD: Index of accessibility to radial highways (9 unique)
    - TAX: Full-value property-tax rate per $10,000 (66 unique)
    - PTRATIO: Pupil-teacher ratio by town (46 unique)
    - B: 1000(Bk - 0.63)^2, where Bk is the proportion of Black people by town (357 unique)
    - LSTAT: Percentage of lower status of the population ( 455 unique)
    - Target Variable: 

MV: Median value of owner-occupied homes in $1000s.

In [ ]:
discrete_vars = ['CHAS'] -- # Only column which has 2 unique value and by column discription considering it as a discrete column
discrete_columns = discrete_vars

In [12]:
# Prepare the data
X = data.drop('MV', axis=1)
y = data['MV']

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Linear Regression

In [14]:
# Step 1: Linear Regression on the dataset
linear_reg = LinearRegression()
linear_reg.fit(X_train, y_train)
y_pred_lr = linear_reg.predict(X_test)
mse_lr = mean_squared_error(y_test, y_pred_lr)
r2_lr = r2_score(y_test, y_pred_lr)

# Normalize the input fields and generating  SGD Regressor

In [15]:
# Step 2: Normalize the input fields and generate a SGD regressor
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
sgd_reg = SGDRegressor(max_iter=1000, tol=1e-3, random_state=42)
sgd_reg.fit(X_train_scaled, y_train)
y_pred_sgd = sgd_reg.predict(X_test_scaled)

mse_sgd = mean_squared_error(y_test, y_pred_sgd)
r2_sgd = r2_score(y_test, y_pred_sgd)


# Feature Selection , Linear Regression and SGD Regressor

In [16]:
# Step 3: Feature selection by dropping high correlation fields
correlation_matrix = X.corr().abs()
upper_triangle = correlation_matrix.where(np.triu(np.ones(correlation_matrix.shape), k=1).astype(bool))

# Drop features with correlation greater than 0.8
high_corr_features = [column for column in upper_triangle.columns if any(upper_triangle[column] > 0.8)]
X_train_dropped = X_train.drop(high_corr_features, axis=1)
X_test_dropped = X_test.drop(high_corr_features, axis=1)

# Linear Regression and SGD Regression after removing high correlated fields
linear_reg.fit(X_train_dropped, y_train)
y_pred_lr_dropped = linear_reg.predict(X_test_dropped)
mse_lr_dropped = mean_squared_error(y_test, y_pred_lr_dropped)
r2_lr_dropped = r2_score(y_test, y_pred_lr_dropped)

scaler_dropped = StandardScaler()
X_train_dropped_scaled = scaler_dropped.fit_transform(X_train_dropped)
X_test_dropped_scaled = scaler_dropped.transform(X_test_dropped)

sgd_reg.fit(X_train_dropped_scaled, y_train)
y_pred_sgd_dropped = sgd_reg.predict(X_test_dropped_scaled)
mse_sgd_dropped = mean_squared_error(y_test, y_pred_sgd_dropped)
r2_sgd_dropped = r2_score(y_test, y_pred_sgd_dropped)

# PCA 

In [17]:
# Step 4: PCA on the inputs and perform regression
pca = PCA(n_components=5)
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

linear_reg.fit(X_train_pca, y_train)
y_pred_lr_pca = linear_reg.predict(X_test_pca)
mse_lr_pca = mean_squared_error(y_test, y_pred_lr_pca)
r2_lr_pca = r2_score(y_test, y_pred_lr_pca)

sgd_reg.fit(X_train_pca, y_train)
y_pred_sgd_pca = sgd_reg.predict(X_test_pca)
mse_sgd_pca = mean_squared_error(y_test, y_pred_sgd_pca)
r2_sgd_pca = r2_score(y_test, y_pred_sgd_pca)

# One - Hot Encoding


In [21]:
# One-hot encoding with ColumnTransformer
column_transformer = ColumnTransformer(
    transformers=[('encoder', OneHotEncoder(handle_unknown='ignore'), discrete_columns)], remainder='passthrough'
)

# Fit the column transformer to the training set
X_train_encoded = column_transformer.fit_transform(X_train)
X_test_encoded = column_transformer.transform(X_test)

# Linear Regression after one-hot encoding
linear_reg.fit(X_train_encoded, y_train)
y_pred_lr_encoded = linear_reg.predict(X_test_encoded)

# Evaluation Metrics
mse_lr_encoded = mean_squared_error(y_test, y_pred_lr_encoded)
r2_lr_encoded = r2_score(y_test, y_pred_lr_encoded)

# Display all results to the user
results = pd.DataFrame({
    "Model": ["Linear Regression", "SGD Regression", "Linear Regression (Dropped Features)", "SGD Regression (Dropped Features)", 
              "Linear Regression (PCA)", "SGD Regression (PCA)", "Linear Regression (One-Hot Encoded)"],
    "MSE": [mse_lr, mse_sgd, mse_lr_dropped, mse_sgd_dropped, mse_lr_pca, mse_sgd_pca, mse_lr_encoded],
    "R-Squared": [r2_lr, r2_sgd, r2_lr_dropped, r2_sgd_dropped, r2_lr_pca, r2_sgd_pca, r2_lr_encoded]
})

print(results)


                                  Model        MSE  R-Squared
0                     Linear Regression  24.291117   0.668760
1                        SGD Regression  24.875080   0.660796
2  Linear Regression (Dropped Features)  25.400905   0.653626
3     SGD Regression (Dropped Features)  25.640159   0.650364
4               Linear Regression (PCA)  30.109735   0.589415
5                  SGD Regression (PCA)  30.254352   0.587443
6   Linear Regression (One-Hot Encoded)  24.291117   0.668760


In [ ]:
X_train, X_test, y_train, y_test

# Optuna Hyperparmater search on KNN, RF, Ridge

In [ ]:
def knn_objective(trial):
    n_neighbors = trial.suggest_int('n_neighbors', 1, 30)
    weights = trial.suggest_categorical('weights', ['uniform', 'distance'])
    p = trial.suggest_int('p', 1, 2)

    model = KNeighborsRegressor(n_neighbors=n_neighbors, weights=weights, p=p)
    score = cross_val_score(model, X_train, y_train, cv=5, scoring='r2').mean()
    return score

def rf_objective(trial):
    n_estimators = trial.suggest_int('n_estimators', 10, 200)
    max_depth = trial.suggest_int('max_depth', 1, 20)
    min_samples_split = trial.suggest_int('min_samples_split', 2, 10)

    model = RandomForestRegressor(n_estimators=n_estimators, max_depth=max_depth, min_samples_split=min_samples_split, random_state=42)
    score = cross_val_score(model, X_train, y_train, cv=5, scoring='r2').mean()
    return score

def ridge_objective(trial):
    alpha = trial.suggest_loguniform('alpha', 1e-3, 10.0)
    solver = trial.suggest_categorical('solver', ['auto', 'svd', 'cholesky', 'lsqr', 'sag'])

    model = Ridge(alpha=alpha, solver=solver, random_state=42)
    score = cross_val_score(model, X_train, y_train, cv=5, scoring='r2').mean()
    return score

In [32]:
# Optimize KNN Regressor
knn_study = optuna.create_study(direction='maximize')
knn_study.optimize(knn_objective, n_trials=50)
print("Best parameters for KNN:", knn_study.best_params)
print("Best R^2 score for KNN:", knn_study.best_value)

# Train and evaluate KNN on test set
knn_model = KNeighborsRegressor(**knn_study.best_params)
knn_model.fit(X_train, y_train)
y_pred_knn = knn_model.predict(X_test)
knn_mse = mean_squared_error(y_test, y_pred_knn)
print("MSE for KNN:", knn_mse)

[I 2024-11-29 21:45:38,795] A new study created in memory with name: no-name-38188f46-44d3-4e96-bd21-842279997fca
[I 2024-11-29 21:45:38,812] Trial 0 finished with value: 0.4274892160676155 and parameters: {'n_neighbors': 2, 'weights': 'distance', 'p': 2}. Best is trial 0 with value: 0.4274892160676155.
[I 2024-11-29 21:45:38,822] Trial 1 finished with value: 0.5103165446659788 and parameters: {'n_neighbors': 21, 'weights': 'distance', 'p': 1}. Best is trial 1 with value: 0.5103165446659788.
[I 2024-11-29 21:45:38,833] Trial 2 finished with value: 0.46286950446185016 and parameters: {'n_neighbors': 9, 'weights': 'uniform', 'p': 2}. Best is trial 1 with value: 0.5103165446659788.
[I 2024-11-29 21:45:38,844] Trial 3 finished with value: 0.5105888212469942 and parameters: {'n_neighbors': 10, 'weights': 'uniform', 'p': 1}. Best is trial 3 with value: 0.5105888212469942.
[I 2024-11-29 21:45:38,856] Trial 4 finished with value: 0.4699322922453543 and parameters: {'n_neighbors': 6, 'weights':

Best parameters for KNN: {'n_neighbors': 5, 'weights': 'distance', 'p': 1}
Best R^2 score for KNN: 0.6149285970110036
MSE for KNN: 20.998923397039146


In [33]:
# Optimize Random Forest Regressor
rf_study = optuna.create_study(direction='maximize')
rf_study.optimize(rf_objective, n_trials=50)
print("Best parameters for Random Forest:", rf_study.best_params)
print("Best R^2 score for Random Forest:", rf_study.best_value)

# Train and evaluate Random Forest on test set
rf_model = RandomForestRegressor(**rf_study.best_params, random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)
rf_mse = mean_squared_error(y_test, y_pred_rf)
print("MSE for Random Forest:", rf_mse)



[I 2024-11-29 21:45:45,814] A new study created in memory with name: no-name-94b6f053-2b7f-4998-8985-6d3588b58a39
[I 2024-11-29 21:45:45,983] Trial 0 finished with value: 0.5406886841549042 and parameters: {'n_estimators': 70, 'max_depth': 1, 'min_samples_split': 10}. Best is trial 0 with value: 0.5406886841549042.
[I 2024-11-29 21:45:46,423] Trial 1 finished with value: 0.5462575950402424 and parameters: {'n_estimators': 175, 'max_depth': 1, 'min_samples_split': 10}. Best is trial 1 with value: 0.5462575950402424.
[I 2024-11-29 21:45:47,487] Trial 2 finished with value: 0.8195392588701113 and parameters: {'n_estimators': 194, 'max_depth': 19, 'min_samples_split': 8}. Best is trial 2 with value: 0.8195392588701113.
[I 2024-11-29 21:45:48,414] Trial 3 finished with value: 0.8218126426009859 and parameters: {'n_estimators': 158, 'max_depth': 10, 'min_samples_split': 3}. Best is trial 3 with value: 0.8218126426009859.
[I 2024-11-29 21:45:48,854] Trial 4 finished with value: 0.800754371228

Best parameters for Random Forest: {'n_estimators': 155, 'max_depth': 14, 'min_samples_split': 3}
Best R^2 score for Random Forest: 0.824349063339878
MSE for Random Forest: 8.410549834140857


In [34]:
# Optimize Ridge Regressor
ridge_study = optuna.create_study(direction='maximize')
ridge_study.optimize(ridge_objective, n_trials=50)
print("Best parameters for Ridge:", ridge_study.best_params)
print("Best R^2 score for Ridge:", ridge_study.best_value)

# Train and evaluate Ridge on test set
ridge_model = Ridge(**ridge_study.best_params, random_state=42)
ridge_model.fit(X_train, y_train)
y_pred_ridge = ridge_model.predict(X_test)
ridge_mse = mean_squared_error(y_test, y_pred_ridge)
print("MSE for Ridge:", ridge_mse)


[I 2024-11-29 21:46:22,303] A new study created in memory with name: no-name-a235ab23-42a9-4fd0-aac0-f0a0a2b09738
C:\Users\GOD\AppData\Local\Temp\ipykernel_17588\1874159057.py:20: FutureWarning: suggest_loguniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use suggest_float(..., log=True) instead.
  alpha = trial.suggest_loguniform('alpha', 1e-3, 10.0)
[I 2024-11-29 21:46:22,314] Trial 0 finished with value: 0.7110607218486449 and parameters: {'alpha': 0.010356765022210486, 'solver': 'lsqr'}. Best is trial 0 with value: 0.7110607218486449.
C:\Users\GOD\AppData\Local\Temp\ipykernel_17588\1874159057.py:20: FutureWarning: suggest_loguniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use suggest_float(..., log=True) instead.
  alpha = trial.suggest_loguniform('alpha', 1e-3, 10.0)
[I 2024-11-29 21:46:22,326] Trial 1 finis

Best parameters for Ridge: {'alpha': 0.0010003432697729975, 'solver': 'cholesky'}
Best R^2 score for Ridge: 0.724433088845696
MSE for Ridge: 24.29117573572531


# Non Hyper Tuned models
            Model                               MSE       R-Squared
    - Linear Regression                       24.291117   0.668760
    - SGD Regression                          24.875080   0.660796
    - Linear Regression (Dropped Features)    25.400905   0.653626
    - SGD Regression (Dropped Features)       25.640159   0.650364
    - Linear Regression (PCA)                 30.109735   0.589415
    - SGD Regression (PCA)                    30.254352   0.587443
    - Linear Regression (One-Hot Encoded)     24.291117   0.668760

# Optuna Optimized Models
## Random Forest 
    Based on the above code the best model seems to be Random Forest with the following result
    Best parameters for Ridge: {'alpha': 0.0010031859321893023, 'solver': 'cholesky'}
    Best R^2 score for Random Forest: 0.8245965893254837
    MSE for Random Forest: 8.046527166338901
## Ridge
    Best parameters for Ridge: {'alpha': 0.0010031859321893023, 'solver': 'cholesky'}
    Best R^2 score for Ridge: 0.7244330813146554
    MSE for Ridge: 24.291175903056494
## KNN
    Best parameters for KNN: {'n_neighbors': 5, 'weights': 'distance', 'p': 1}
    Best R^2 score for KNN: 0.6149285970110036
    MSE for KNN: 20.998923397039146



In [ ]:
# Compare models and save the best one
best_model = None
best_score = -np.inf

# Check KNN
if knn_study.best_value > best_score:
    best_model = knn_model
    best_score = knn_study.best_value
    best_model_name = 'KNN'
    best_mse = knn_mse

# Check Random Forest
if rf_study.best_value > best_score:
    best_model = rf_model
    best_score = rf_study.best_value
    best_model_name = 'Random Forest'
    best_mse = rf_mse

# Check Ridge
if ridge_study.best_value > best_score:
    best_model = ridge_model
    best_score = ridge_study.best_value
    best_model_name = 'Ridge'
    best_mse = ridge_mse

# Save the best model
print(f"Best model is {best_model_name} with R^2 score: {best_score} and MSE: {best_mse}")
joblib.dump(best_model, f'model/best_model_{best_model_name.lower()}.pkl')


Best model is Random Forest with R^2 score: 0.824349063339878 and MSE: 8.410549834140857


['best_model_random forest.pkl']

### For Flask code please refer
```py
import optuna
import joblib
from flask import Flask, request, render_template, redirect, url_for
import pandas as pd
import numpy as np

model = joblib.load('model/best_model_random_forest.pkl')

columns = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PT', 'B', 'LSTAT']

# Create Flask app
app = Flask(__name__)

@app.route('/')
def home():
    return render_template('index.html')


@app.route('/predict', methods=['POST'])
def predict():
    try:
        input_format = request.form.get("inputFormat")
        if input_format == "json":
            
            json_data = request.form["jsonData"]
            input_data = pd.read_json(json_data)
        else:
            
            input_data = [
                float(request.form[column]) for column in columns
            ]
            input_data = pd.DataFrame([input_data], columns=columns)
        
     
        predictions = model.predict(input_data)
        prediction_texts = [f'Predicted Median Value of Owner-Occupied Homes: ${pred:.2f}' for pred in predictions]
        return render_template('result.html', prediction_text=prediction_texts)
    except Exception as e:
        return render_template('result.html', prediction_text=[f'Error occurred: {str(e)}'])

@app.route('/result')
def prediction_result():
    prediction_text = request.args.get('prediction_text')
    return render_template('result.html', prediction_text=prediction_text)

if __name__ == '__main__':
    app.run(debug=True)
```

#### For Frontend Webpage - Index.html
```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Boston Housing Price Prediction</title>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css">
</head>
<body>
    <div class="container">
        <h2 class="text-center mt-4">Boston Housing Price Prediction</h2>
        <form action="/predict" method="post">
            <div class="form-group">
                <label for="inputFormat">Input Format:</label>
                <select class="form-control" id="inputFormat" name="inputFormat" onchange="toggleInputFields()">
                    <option value="manual">Manual Entry</option>
                    <option value="json">JSON Input</option>
                </select>
            </div>

            <div id="manualInput">
                <div class="form-group">
                    <label for="CRIM">CRIM (Per capita crime rate by town):</label>
                    <input type="number" step="0.01" class="form-control" id="CRIM" name="CRIM">
                </div>
                <div class="form-group">
                    <label for="ZN">ZN (Proportion of residential land zoned for lots over 25,000 sq. ft.):</label>
                    <input type="number" step="0.01" class="form-control" id="ZN" name="ZN">
                </div>
                <div class="form-group">
                    <label for="INDUS">INDUS (Proportion of non-retail business acres per town):</label>
                    <input type="number" step="0.01" class="form-control" id="INDUS" name="INDUS">
                </div>
                <div class="form-group">
                    <label for="CHAS">CHAS (Charles River dummy variable [1 if tract bounds river; 0 otherwise]):</label>
                    <input type="number" step="1" class="form-control" id="CHAS" name="CHAS">
                </div>
                <div class="form-group">
                    <label for="NOX">NOX (Nitric oxide concentration [parts per 10 million]):</label>
                    <input type="number" step="0.001" class="form-control" id="NOX" name="NOX">
                </div>
                <div class="form-group">
                    <label for="RM">RM (Average number of rooms per dwelling):</label>
                    <input type="number" step="0.01" class="form-control" id="RM" name="RM">
                </div>
                <div class="form-group">
                    <label for="AGE">AGE (Proportion of owner-occupied units built before 1940):</label>
                    <input type="number" step="0.01" class="form-control" id="AGE" name="AGE">
                </div>
                <div class="form-group">
                    <label for="DIS">DIS (Weighted distances to five Boston employment centers):</label>
                    <input type="number" step="0.01" class="form-control" id="DIS" name="DIS">
                </div>
                <div class="form-group">
                    <label for="RAD">RAD (Index of accessibility to radial highways):</label>
                    <input type="number" step="1" class="form-control" id="RAD" name="RAD">
                </div>
                <div class="form-group">
                    <label for="TAX">TAX (Full-value property-tax rate per $10,000):</label>
                    <input type="number" step="1" class="form-control" id="TAX" name="TAX">
                </div>
                <div class="form-group">
                    <label for="PTRATIO">PTRATIO (Pupil-teacher ratio by town):</label>
                    <input type="number" step="0.01" class="form-control" id="PTRATIO" name="PTRATIO">
                </div>
                <div class="form-group">
                    <label for="B">B (1000(Bk - 0.63)^2, where Bk is the proportion of Black people by town):</label>
                    <input type="number" step="0.01" class="form-control" id="B" name="B">
                </div>
                <div class="form-group">
                    <label for="LSTAT">LSTAT (Percentage of lower status of the population):</label>
                    <input type="number" step="0.01" class="form-control" id="LSTAT" name="LSTAT">
                </div>
            </div>

            <div id="jsonInput" style="display: none;">
                <div class="form-group">
                    <label for="jsonData">Enter Input Data in JSON Format (for multiple rows, use an array of JSON objects):</label>
                    <textarea class="form-control" id="jsonData" name="jsonData" rows="10"></textarea>
                </div>
            </div>

            <button type="submit" class="btn btn-primary">Predict</button>
        </form>
    </div>

    <div class="container text-center mt-5">
        <a href="/" class="btn btn-secondary">Predict Another Value</a>
    </div>

    <script>
        function toggleInputFields() {
            var inputFormat = document.getElementById("inputFormat").value;
            if (inputFormat === "manual") {
                document.getElementById("manualInput").style.display = "block";
                document.getElementById("jsonInput").style.display = "none";
            } else {
                document.getElementById("manualInput").style.display = "none";
                document.getElementById("jsonInput").style.display = "block";
            }
        }
    </script>
</body>
</html>
```

#### For Frontend Webpage - Result.html
```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Prediction Result</title>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css">
</head>
<body>
    <div class="container">
        <h2 class="text-center mt-4">Prediction Results</h2>
        <div class="mt-4">
            {% if prediction_text %}
                <ul class="list-group">
                    {% for text in prediction_text %}
                        <li class="list-group-item text-success">{{ text }}</li>
                    {% endfor %}
                </ul>
            {% endif %}
        </div>
        <div class="mt-4 text-center">
            <a href="/" class="btn btn-primary">Predict Another Value</a>
        </div>
    </div>
</body>
</html>
```

#### For the Flask result please Refer the Image folder or use the below as reference
![Maunal Entry](images/Manual_Entry.png)
![Single Json](images/Single_json.png)
![Multi Json](images/Multi_json.png)
![Result](images/Result.png)